genet/core.py

```diff
@@ -686,14 +691,18 @@ def add_nodes(self, nodes_and_attribs: dict, silent: bool = False, ignore_change
         reduce changelog bloat.
         :return:
         """
-        # check for clashing nodes
+        # check for spatial info
+        for node_id, value in nodes_and_attribs.items():
+            if ('lat' not in value.keys() and 'lon' not in value.keys()) and \
+                    ('x' not in value.keys() and 'y' not in value.keys()):
+                raise RuntimeError(f'Cannot add Node `{node_id}` without any spatial information')
+
+        # check for clashing node IDs
```

Comment:
Duplicated code 🙅 `add_node` is nothing special, so should delegate to `add_nodes`. Less code means less code to test and maintain and less code to go wrong.